File: playground.js
const { spawn, exec } = require("node:child_process");
// const { stdin, stdout, stderr } = require("node:process");

// stdin.on("data", (data) => {
//   console.log("You wrote: ", data.toString());
// });

// stdout.write("Hello, what's your name? ");
// stderr.write("Error: ");

// echo "something string" | tr ' ' '\n'

const subprocess = spawn(
  "./playground",
  ["some string", "-f", 34, "some more string", "-u"],
  {
    env: { MODE: "development" },
  }
);

subprocess.stdout.on("data", (data) => {
  console.log("Output from the C application: ", data.toString("utf-8"));
});

subprocess.stderr.on("data", (data) => {
  console.log("Error from the C application: ", data.toString("utf-8"));
});

subprocess.stdin.write("Some text on stdin from Node.js application.");
subprocess.stdin.on("error", (error) => {
  console.log("Error from the Node.js application: ", error.toString());
});
subprocess.stdin.end();

// exec(
//   " source .bashrc && echo 'something string' | tr ' ' '\n'",
//   {
//     shell: "/bin/zsh",
//   },
//   (error, stdout, stderr) => {
//     if (error) {
//       console.error(error);
//       return;
//     }

//     console.log(stdout);

//     console.log(`stderr: ${stderr}`);
//   }
// );

File: playground.c
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>

int main(int argc, char *argv[]) {

//   for (int i = 0; i < argc; i++) {
//       printf("argv number %d is: %s\n", i, argv[i]);
//   }

//   printf("Process ID: %d\n", getpid());
//   printf("Parent Process ID: %d\n", getppid());

//   printf("Mode is: %s\n", getenv("MODE"));

  fprintf(stdout, "(Comming from C) Some text on stdout.\n");
  fprintf(stderr, "(Comming from C) Some text on stderr.\n");

  int c = fgetc(stdin);
  while (c != EOF) {
    fprintf(stdout, "%c", c);
    fflush(stdout);
    c = fgetc(stdin);
  }

  return 0;
}